homeassistant/components/bluetooth/manager.py

```diff
@@ -483,15 +483,15 @@ def _async_remove_callback() -> None:
         # immediately with the last packet so the subscriber can see the
         # device.
         all_history = self._get_history_by_type(connectable)
-        if (
-            (address := callback_matcher.get(ADDRESS))
-            and (service_info := all_history.get(address))
-            and ble_device_matches(callback_matcher, service_info)
-        ):
-            try:
-                callback(service_info, BluetoothChange.ADVERTISEMENT)
-            except Exception:  # pylint: disable=broad-except
-                _LOGGER.exception("Error in bluetooth callback")
+        for service_info in all_history.values():
```

Comment:
This can be hundreds of devices and if they have 20-30 bluetooth integrations this be a drag on startup.

I think we can optimize this to a bit to use the existing path if there the matcher is only ADDRESS + optional connectable which will be most cases, and if there are more fields in the matcher use this new path

Reply:
I think the added commit should optimize as you suggested.